core/src/main/java/io/grpc/internal/PickFirstLeafLoadBalancer.java

```diff
@@ -0,0 +1,475 @@
+        // the current picker in-place. But ignoring the potential optimization is simpler.
+        updateBalancingState(CONNECTING, new Picker(PickResult.withNoResult()));
+        break;
+      case READY:
+        updateBalancingState(READY, new Picker(PickResult.withSubchannel(subchannel)));
+        shutdownRemaining(subchannel);
+        break;
+      case TRANSIENT_FAILURE:
+        // If we are looking at current channel, request a connection if possible
+        if (addressIndex.isValid()
+            && subchannels.get(addressIndex.getCurrentAddress()).getSubchannel() == subchannel) {
+          addressIndex.increment();
+          requestConnection();
+
+          // If no addresses remaining, go into TRANSIENT_FAILURE
+          if (!addressIndex.isValid()) {
+            helper.refreshNameResolution();
+            updateBalancingState(TRANSIENT_FAILURE,
+                new Picker(PickResult.withError(stateInfo.getStatus())));
+          }
+        }
+        break;
+      default:
+        throw new IllegalArgumentException("Unsupported state:" + newState);
+    }
+  }
+
+  private void updateBalancingState(ConnectivityState state, SubchannelPicker picker) {
+    if (state != currentState || state == READY || state == TRANSIENT_FAILURE) {
+      currentState = state;
+      helper.updateBalancingState(state, picker);
+    }
+  }
+
+  @Override
+  public void shutdown() {
+    for (SubchannelData subchannelData : subchannels.values()) {
+      subchannelData.getSubchannel().shutdown();
+    }
+    subchannels.clear();
+  }
+
+  /**
+  * Shuts down remaining subchannels. Called when a subchannel becomes ready, which means
+  * that all other subchannels must be shutdown.
+  */
+  private void shutdownRemaining(Subchannel activeSubchannel) {
+    for (SubchannelData subchannelData : subchannels.values()) {
+      if (!subchannelData.getSubchannel().equals(activeSubchannel)) {
+        subchannelData.getSubchannel().shutdown();
+      }
+    }
+    subchannels.clear();
+    subchannels.put(getAddress(activeSubchannel), new SubchannelData(activeSubchannel, READY));
+  }
+
+  /**
+  * Requests a connection to the next applicable address' subchannel, creating one if necessary
+  * If the current channel has already attempted a connection, we attempt a connection
+  * to the next address/subchannel in our list.
+  */
+  @Override
+  public void requestConnection() {
+    if (subchannels.size() == 0) {
+      return;
+    }
+    if (addressIndex.isValid()) {
+      Subchannel subchannel = subchannels.containsKey(addressIndex.getCurrentAddress())
+          ? subchannels.get(addressIndex.getCurrentAddress()).getSubchannel()
+          : createNewSubchannel(addressIndex.getCurrentAddress());
+
+      ConnectivityState subchannelState =
+          subchannels.get(addressIndex.getCurrentAddress()).getState();
+      if (subchannelState == IDLE) {
+        subchannel.requestConnection();
+      } else {
```

Comment:
this should be 
```java
} else if (subchannelState == TRANSIENT_FAULRE) {
addressIndex.increment();
requestConnection();
}
```

this means we should skip request connection when in READY and CONNECTING state. But first when in READY, we should put the addressIndex at the active subchannel.

Reply:
Since we update a subchannel's state to CONNECTING instead of keeping it in TRANSIENT_FAILURE when it comes out of backoff, we would want it to iterate when CONNECTING as well since we don't distinguish between CONNECTING when iterating and CONNECTING when coming out of backoff. The consequence of this is that when we have an address update in the following case:

1: TF
2: CONNECTING (current)
3: IDLE

We start another connection at the new address instead of letting at address 2 complete before:

1: TF
2: CONNECTING 
4: CONNECTING (current)

If in READY, requestConnection() is never called (at least internally). But, I will add a check for this.

Reply:
`requestConnection()` can be called directly on the LB externally from the channel, even it is in READY.

I see the point that CONNECTING could be coming out of backoff. I imagine either first time CONNECTING or CONNECTING out of backoff can be treated the same, as it would soon return a result(TF/Ready) and we can move on, similar to what we do the initial iteration. But we if we do aggressively iteration regardless, may not be a problem.
But there are things we can do to distinguish whether out of backoff: to let the subchannel's state being TF until it gets READY. Then here we can do incrementing only in TF, because then we know CONNECTING means first time CONNECTING, or became READY before.
The CONNECTING issue is subtle here. It is fine to change in a follow up PR.

Reply:
So that's what we used to do--we would leave the subchannel's state as TF until it became READY in anticipation for this case. But, you did make a valid point that we should not do hacky stuff. If we did that, it would kind of implement sticky TF twice--once in the subchannel layer and once in the LB layer, if that makes sense. 

I personally think the current behavior is okay, especially since pursuing the other option would either make us add additional logic that might make the code more dense or mutate the states in a way that gets more confusing than it is.

Reply:
Note that it is still different from what we used to do. The subchannel was in sticky TF only when overall state is in TF, but here we are subchannel can be implemented sticky TF regardless of the overall state.